### DISCUSSION TITLE
More precise return type of handleSubmit

### DISCUSSION THREAD

**Discussion by JonasDoe** [react-hook-form/react-hook-form#10965] [Q&A]
**Is your feature request related to a problem? Please describe.**
At the moment, when I do something like
```typescript
const { handleSubmit } = useForm();
const handleData = (data) => sendData().catch(console.log);
return <form onSubmit={handleSubmit(handleData)}>...</form>
```
`@typescript-eslint/no-misused-promises` complains that: `"Promise-returning function provided to attribute where a void return was expected."`. But since my function doesn't return a `Promise`, that's not actually not the case here. The assumption comes from the type `UseFormHandleSubmit`, which returns a function with its result being just `Promise<void>` anyway.

**Describe the solution you'd like**
It would be great if the return value of `FormSubmitHandler` would be able to infer from the passed `SubmitHandler` whether the created callback actually returns a `Promise` or not.

**Describe alternatives you've considered**
I either have to disable the eslint rule for that line, or opt for a verbose solution as
```typescript
onSubmit={(event) => void form.handleSubmit(onSubmit)(event)}
```

**Comment by Moshyfawn** [comment]
https://github.com/orgs/react-hook-form/discussions/9325#discussioncomment-4060566

**Comment by JonasDoe** [comment]
Thanks for the fast response! This feature sounds helpful, but I'm still wondering whether that's the best way to settle the issue. If it was possible that `react-hook-form` infered the return type properly, this would be more precise and probably less error prone. I mean, I my case the issue is not that I don't know how to to ignore a `Promise`, but that there's no `Promise` at all.
But if that's too hard to implement, nevermind. I'm not entitled to it anyway. :)

**Comment by vnugent** [comment]
I'm facing this same issue (originally posted [here](https://github.com/orgs/react-hook-form/discussions/10855)).

**Comment by crispyricepc** [comment]
FWIW, [@typescript-eslint/no-misused-promises](https://typescript-eslint.io/rules/no-misused-promises/#checksvoidreturn) has rule exceptions setup specifically for JSX void promises like this, so disabling the rule for cases where a void promise is passed as an attribute to a JSX element (like this) can "fix" the problem.

```jsonc
"rules": {
  // ...
  "@typescript-eslint/no-misused-promises": ["error", { "checksVoidReturn": { "attributes": false } }],
}
```

**Comment by jpenna** [comment]
Disabling the linting rule can affect other parts of the code, not just this one related to `handleSubmit`. 

What is the reason for returning a `Promise` from `handleSubmit` if it isn't used? Are there cases where it is awaited?

**Comment by EnzoPV25** [comment]
Any update on this ? 2 years later facing the same isssue
